onnxscript/optimizer/constant_folding_ir.py

```diff
@@ -0,0 +1,718 @@
+
+        # Do incremental shape inference
+        if self._do_shape_inference and not is_control_flow_op(node):
+            self._do_inference(node)
+
+        if node.domain not in self.opset_imports:
+            return None
+        version = self.opset_imports[node.domain]
+        op_optimizers = registry.lookup_evaluators(node.domain, node.op_type, version)
+        for optimizer in op_optimizers:
+            assert optimizer
+            context = orp.RewriterContext()
+            output = optimizer(context, node)
+            if output is not None:
+                if isinstance(output, Replacement):
+                    return output
+                if isinstance(output, ir.Value):
+                    output = [output]
+                return Replacement(output, context.nodes)
+
+        if is_control_flow_op(node) or is_non_deterministic_op(node):
+            return None
+
+        if any((x is not None and x.const_value is None) for x in node.inputs):
+            return None
+
+        input_values = [x.const_value.numpy() if x is not None else None for x in node.inputs]
+
+        # Filter out bfloat16 cases?
+        def convert(av):
+            if isinstance(av, ir.AttrTensor):
+                return serde.serialize_tensor(av.value)
+            return av.value
+
+        attr_values = {name: convert(attr) for name, attr in node.attributes.items()}
+        outputs = reference_evaluator.evaluate(
+            node.domain, node.op_type, version, *input_values, **attr_values
+        )
+
+        if outputs is None:
+            return None
+        if len(node.outputs) == 1 and not isinstance(outputs, (tuple, list)):
+            replacement = self.new_constant(node.outputs[0], outputs)
+            if is_constant_op(node) or replacement is None:
+                return None
+            # self.add_count(op, outputs.size)
+            return Replacement(replacement.outputs, [replacement])
+        else:
+            logger.warning(
+                "Skipping constant folding for op %s with multiple outputs.", node.op_type
+            )
+        return None
+
+    def replace_node(self, node: ir.Node, replacement, root: ir.Graph | ir.Function):
+        # TODO: apply delta! what about opset_imports?
+        old_values = node.outputs
+        new_values = replacement.new_outputs
+        for old_value, new_value in zip(old_values, new_values):
+            # Propagate relevant info from old value to new value
+            # TODO(Rama): Perhaps we should merge old and new types. As of now, new
+            # values don't have type information. Note that this could be a problem
+            # for semantics-altering rewrite-rules: we should allow users to override
+            # this for such rules.
+            new_value.type = old_value.type
+            new_value.shape = old_value.shape
+            new_value.const_value = old_value.const_value
+            new_value.name = old_value.name
```

Comment:
We can create a copy() method for nodes

Reply:
yes, at least, we need some common utility we can call here.